onnx/defs/schema.h

```diff
@@ -655,14 +662,21 @@ class OpSchemaRegistry final {
   }
 };
 
-#define ONNX_OPERATOR_SCHEMA(name) \
-  ONNX_OPERATOR_SCHEMA_UNIQ_HELPER(__COUNTER__, name)
-#define ONNX_OPERATOR_SCHEMA_UNIQ_HELPER(Counter, name) \
-  ONNX_OPERATOR_SCHEMA_UNIQ(Counter, name)
-#define ONNX_OPERATOR_SCHEMA_UNIQ(Counter, name)                 \
+// ONNX_OPERATOR_SCHEMA is implemented either with static registration or by exposing
+// a factory method.  
+#ifndef ONNX_NON_STATIC_SCHEMA_INITIALIZATION
+#define ONNX_OPERATOR_SCHEMA(name, domain, ver, impl) \
   static ONNX_NAMESPACE::OpSchemaRegistry::OpSchemaRegisterOnce( \
-      op_schema_register_once##name##Counter) =                  \
-      OpSchema(#name, __FILE__, __LINE__)
+      op_schema_register_once_##name##_##domain##_ver##ver) = impl\
+      .SetName(#name).SetDomain(domain##_NAME).SinceVersion(ver).SetLocation(__FILE__, __LINE__)
+#else
+template <typename T> OpSchema GetOpSchema();
+#define ONNX_OPERATOR_SCHEMA(name, domain, ver, impl) \
+    class name##_##domain##_ver##ver##_OpSchemaId{}; \
+    template<> OpSchema ONNX_NAMESPACE::GetOpSchema< name##_##domain##_ver##ver##_OpSchemaId>(){ \

```

Comment:
Actually could you elaborate how does it work? Where is this `GetOpSchema` declared and how is user code supposed to call it?

Reply:
GetOpSchema is declared slightly above this.  Macro call sites define its specializations.  

Based on the feedback here, I'm preparing a change to remove the original mode of registration during static initialization.  This will include a new file with related declarations for each schema, and functions which retrieve and/or register the schema of specific operator sets.